Makefile

```diff
@@ -35,22 +35,38 @@ LDFLAGS := "-X main.version=$(VERSION) \
 .PHONY: build
 build: manifests generate local-gadget kubectl-gadget gadget-container
 
-# local-gadget
-.PHONY: local-gadget
-local-gadget:
-	make -C gadget-container ebpf-objects
-	go build \
-		-tags withebpf \
-		-ldflags "-X main.version=$(VERSION)" \
-		-o local-gadget \
-		github.com/kinvolk/inspektor-gadget/cmd/local-gadget
-
 # make does not allow implicit rules (with '%') to be phony so let's use
 # the 'phony_explicit' dependency to make implicit rules inherit the phony
 # attribute
 .PHONY: phony_explicit
 phony_explicit:
 
+# local-gadget
+
+LOCAL_GADGET_TARGETS = \
+	local-gadget-linux-amd64
+	# local-gadget-linux-arm64 # TODO: local-gadget does not work on arm64 yet
```

Comment:
Out of curiosity:
> local-gadget does not work on arm64 yet

do you have a tracking issue to see what's going on ?

Reply:
I didn't file an issue. I had this error in the socket-collector gadget:
```
make[2]: Entering directory '/home/alban/go/src/github.com/kinvolk/inspektor-gadget/pkg/gadgets/socket-collector/tracer/bpf'
GO111MODULE=on CGO_ENABLED=1 GOOS=linux go generate ../
fork/exec /tmp/go-build1226201060/b001/exe/bpf2go: exec format error
```

It seems like something to fix in the cross-compiling: the intermediary tool `bpf2go` used in `go generate` should not be cross-compiled but compiled normally. Maybe something to fix with the environment variables `GOARCH` and `GOHOSTARCH`.

https://github.com/kinvolk/inspektor-gadget/blob/38680fe3ae34f9d14a965670d294b00dfe8b87bd/pkg/gadgets/socket-collector/tracer/tracer.go#L30-L31

But we wanted to make a release today, so I was thinking we could get amd64 work in today's release, and arm64 later when we have time to look closer at the issue.

Reply:
Right, looking at the `ebpf2go` code `-target bpfel` should be enough. `GOARCH` should be set / unset between the generation steps.

Reply:
The problem in the socket-collector can be fixed with the following:
```diff
--- a/pkg/gadgets/socket-collector/tracer/tracer.go
+++ b/pkg/gadgets/socket-collector/tracer/tracer.go
@@ -27,8 +27,8 @@ import (
        "github.com/vishvananda/netns"
 )
 
-//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterTCPv4 ./bpf/tcp4-collector.c -- -I../../.. -Werror -O2 -g -c -x c
-//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterUDPv4 ./bpf/udp4-collector.c -- -I../../.. -Werror -O2 -g -c -x c
+//go:generate sh -c "GOOS='linux' GOARCH='amd64' go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterTCPv4 ./bpf/tcp4-collector.c -- -I../../.. -Werror -O2 -g -c -x c"
+//go:generate sh -c "GOOS='linux' GOARCH='amd64' go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterUDPv4 ./bpf/udp4-collector.c -- -I../../.. -Werror -O2 -g -c -x c"
 
 func parseIPv4(ipU32 uint32) string {
        ipBytes := make([]byte, 4)
```

Then, the build of local-gadget needs to have:
```
export CGO_ENABLED=1
export CC=$${GOARCH%64}-$${GOOS}-gnu-gcc
```
because it uses cgo, libseccomp, etc., we need to have a full cross compiler installation.

* It seems that Fedora only offer partial support for cross compilation: they only support cross compilation of the kernel but not of all libraries.
* Debian offers support for cross compilation, so I started with the following Docker build based on a Debian container:
```
FROM golang:1.17 AS builder

RUN \
	dpkg --add-architecture arm64 && \
	apt-get update && \
	apt-get install -y gcc-aarch64-linux-gnu build-essential crossbuild-essential-arm64 && \
	apt-get install -y libseccomp2:arm64 libseccomp-dev:arm64

COPY go.mod go.sum /cache/
RUN cd /cache && go mod download

ADD . /go/src/github.com/kinvolk/inspektor-gadget
WORKDIR /go/src/github.com/kinvolk/inspektor-gadget

RUN \
	CGO_ENABLED=1 \
	CC=aarch64-linux-gnu-gcc \
	GOOS=linux GOARCH=arm64 \
	PKG_CONFIG_PATH=/usr/lib/aarch64-linux-gnu/pkgconfig \
	go build \
	-tags withebpf \
	-o local-gadget-linux-arm64 \
	github.com/kinvolk/inspektor-gadget/cmd/local-gadget
```
and compile with `docker build -f local-gadget.Dockerfile .`.

This produces the following:
```
# file local-gadget-linux-arm64 
local-gadget-linux-arm64: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=df5835d20955d90bc44fa3ff01734190fe3f4306, for GNU/Linux 3.7.0, not stripped
```